### ISSUE TITLE
[AArch64] Generate sqdmlal

### ISSUE THREAD

**Issue by sjoerdmeijer** [llvm/llvm-project#49997]
|  |  |
| --- | --- |
| Bugzilla Link | [50653](https://llvm.org/bz50653) |
| Version | trunk |
| OS | Linux |
| CC | @Arnaud-de-Grandmaison-ARM,@DMG862,@smithp35 |

## Extended Description 
Raising this missed optimisation opportunity in case someone finds this interesting.

For this input:

#include "arm_neon.h"
int32_t
t_vqdmlalh_lane_s16 (int32_t a, int16_t b, int16x4_t c) {
  return vqdmlalh_lane_s16 (a, b, c, 0);
}

We are not generating this multiply-accumulate variant that gcc generates:

t_vqdmlalh_lane_s16:
        dup     v2.4h, w1
        fmov    s1, w0
        sqdmlal s1, h2, v0.h[0]
        fmov    w0, s1
        ret

We get this instead:

t_vqdmlalh_lane_s16:                    // @&#8203;t_vqdmlalh_lane_s16
        fmov    s1, w1
        sqdmull v0.4s, v1.4h, v0.4h
        fmov    s1, w0
        sqadd   s0, s1, s0
        fmov    w0, s0
        ret

See also https://godbolt.org/z/41nMxM5q1

**Comment by overmighty** [comment]
I am looking into this.

There are equivalent missed optimization opportunities with the following intrinsics too:

- `vqdmlalh_s16`
- `vqdmlslh_s16`
- `vqdmlalh_laneq_s16`
- `vqdmlslh_lane_s16`
- `vqdmlslh_laneq_s16`

**Comment by overmighty** [comment]
I found a fix for this issue.

I also found that `sqdmlal`/`sqdmlsl` instructions are in fact generated from the following intrinsics, as long as the lane number passed is not 0:

- `vqdmlalh_lane_s16`
- `vqdmlalh_laneq_s16`
- `vqdmlslh_lane_s16`
- `vqdmlslh_laneq_s16`

For example, for this C code:

```c
int32_t u_vqdmlalh_lane_s16(int32_t a, int16_t b, int16x4_t v) {
    return vqdmlalh_lane_s16(a, b, v, 1);
}
```

Clang generates this AArch64 assembly code:

```asm
u_vqdmlalh_lane_s16:                    // @u_vqdmlalh_lane_s16
        fmov    s1, w1
        fmov    s2, w0
        sqdmlal v2.4s, v1.4h, v0.h[1]
        fmov    w0, s2
        ret
```

See https://godbolt.org/z/sjfvKoK81 for all my experiments.

The different kind of DAGs generated when the lane number is not 0 is matched by this TableGen definition:

https://github.com/llvm/llvm-project/blob/f8d976171f2a1b7bf9268929f77904973edb0378/llvm/lib/Target/AArch64/AArch64InstrFormats.td#L8850-L8862

I can simply remove this definition and the `v1i32_indexed` instructions will be used instead because of my fix, however it will result in an extra `dup` instruction:

```asm
u_vqdmlalh_lane_s16:                    // @u_vqdmlalh_lane_s16
	fmov	s1, w0
	fmov	s2, w1
	dup	v0.4h, v0.h[1]
	sqdmlal	s1, s2, v0.h[0]
	fmov	w0, s1
	ret
```

Therefore I would like to know if that FIXME comment really is an issue or not.

Something else I found is that several TableGen definitions seem to be unused, because they use the `FPR16` class of AArch64 registers, however they are part of a multiclass that is only invoked to declare instructions that take integer operands, and this register class can only be used for `f16` and `bf16` floats.

One example is this `def i16` here:

https://github.com/llvm/llvm-project/blob/f8d976171f2a1b7bf9268929f77904973edb0378/llvm/lib/Target/AArch64/AArch64InstrFormats.td#L7007-L7013

`SIMDThreeScalarMixedTiedHS` is only invoked here, for scalar `SQDMLAL` and `SQDMLSL` which take integer operands:

https://github.com/llvm/llvm-project/blob/f8d976171f2a1b7bf9268929f77904973edb0378/llvm/lib/Target/AArch64/AArch64InstrInfo.td#L4954-L4955

Should something be done about these definitions?
